### PR TITLE
[release-3.1] cherry pick: Generate flamegraph at runtime  (#5697)

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: Yang Keao <keao.yang@yahoo.com>

###  What have you changed?

cherry-pick: #5697  #5933  #6064 

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

They have just been tested manually :(

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Refer to a related PR or issue link

#5666 

###  Any examples?

Assume that tikv status_server is listening at localhost:8080

```
curl "http://localhost:8080/debug/pprof/profile?frequency=50&seconds=10" # this will start cpu sampler at 10/sec frequency and continues 10 secs.
```

```
curl "http://localhost:8080/debug/pprof/profile?frequency=50&seconds=10" --header  "Content-Type: application/protobuf" # this will return a pprof protobuf
```

### Notice:

The implementation of `rsperftools` is not efficient ([some performance improvement](https://github.com/YangKeao/rsperftools/issues) will land these days). So too high frequency will make it hold lock all the time and may lead to long response time and high cpu usage.

### Panic:

Under these situations, `rsperftools` will crash or deadlock:

- profiling tick happens when program is propagating thrown exception

flamegraph

[flamegraph.svg.gz](https://github.com/tikv/tikv/files/3797699/flamegraph.svg.gz)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -124,6 +125,8 @@ match_template = { path = "components/match_template" }
 cop_datatype = { path = "components/cop_datatype" }
 cop_codegen = { path = "components/cop_codegen" }
 codec = { path = "components/codec" }
+pprof = { version = "0.3", features = ["flamegraph", "protobuf"] }
```
Comment (breeswish) [inline Cargo.toml]: Duplicated dependency?

File: src/server/status_server.rs [hunk 2]
```diff
@@ -41,30 +50,205 @@ impl StatusServer {
...
+    {
+        Response::builder()
+            .status(status_code)
+            .body(message.into())
+            .unwrap()
+    }
 
-        // TODO: support TLS for the status server.
-        let builder = Server::try_bind(&addr)?;
+    fn extract_thread_name(thread_name: &str) -> String {
```
Comment (breeswish) [inline src/server/status_server.rs]: Missing `#[cfg(target_os = "linux")]`

Comment (YangKeao) [reply src/server/status_server.rs]: Fixed!

File: Cargo.toml (hunk 1 above)
Comment (YangKeao) [reply Cargo.toml]: Fixed

**Comment by AndreMouche** [comment]
friendly ping @YangKeao

**Pull request merged**